JDK-8318636: Add jcmd to print annotated process memory map

tstuefe opened the pull request:
Analysts and supporters often use /proc/xx/maps to make sense of the memory footprint of a process.

Interpreting the memory map correctly can help when used as a complement to other tools (e.g. NMT). There even exist tools out there that attempt to annotate the process memory map with JVM information.

That, however, can be more accurately done from within the JVM, at least for mappings originating from hotspot. After all, we can correlate the mapping information in NMT with the VMA map.

Example output (from a spring petstore run):

[example_system_map.txt](https://github.com/openjdk/jdk/files/13179054/example_system_map.txt)

This patch adds the VM annotations for VMAs that are *mmaped*. I also have an experimental patch that works with malloc'ed memory, but it is not ready yet for consumption and I leave that part of for now.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318636](https://bugs.openjdk.org/browse/JDK-8318636): Add jcmd to print annotated process memory map (**Enhancement** - P4)


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [3829abbf](https://git.openjdk.org/jdk/pull/16301/files/3829abbfad65492d7682a17107f33e86719fc9b5)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16301/head:pull/16301` \
`$ git checkout pull/16301`

Update a local copy of the PR: \
`$ git checkout pull/16301` \
`$ git pull https://git.openjdk.org/jdk.git pull/16301/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16301`

View PR using the GUI difftool: \
`$ git pr show -t 16301`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16301.diff">https://git.openjdk.org/jdk/pull/16301.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16301#issuecomment-1774178871)

plummercj commented:
The hs_err file includes a dump of the maps file. It seems these annotations would be useful there also.

tstuefe commented:
> The hs_err file includes a dump of the maps file. It seems these annotations would be useful there also.

I thought about this too. The problem is the printing may take long since it is O^2 (iterate through VMAs, for each one iterate NMT regions). We had cases where the process ran with >20 mio VMAs.

But thinking further, this is a problem anyway, since the hs-err file was >2GB. And we do have per-step timeouts in the hs-err writer. So, stopping after a set number of VMAs (lets say 100k, that covers 99% of all processes) and otherwise relying on the hs-err timeout code may be ok.

I'll give it a shot.

tstuefe commented:
@plummercj on second thought, I'd rather leave the hs-err part for a follow-up RFE.

zhengyu123 commented:
> > The hs_err file includes a dump of the maps file. It seems these annotations would be useful there also.
> 
> I thought about this too. The problem is the printing may take long since it is O^2 (iterate through VMAs, for each one iterate NMT regions). We had cases where the process ran with >20 mio VMAs (ZGC with a 5TB heap).
> 
> But thinking further, this is a problem anyway, since the hs-err file was >2GB. And we do have per-step timeouts in the hs-err writer. So, stopping after a set number of VMAs (lets say 100k, that covers 99% of all processes) and otherwise relying on the hs-err timeout code may be ok.
> 
> I'll give it a shot.

IIRC, both lists are sorted, should be a way to reduce time complexity.

jdksjolen commented on test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTest.java:
```diff
@@ -0,0 +1,66 @@
+        OutputAnalyzer output = executor.execute("System.map");
+        output.reportDiagnosticSummary();
+        // Only with NMT we get VM annotations
+        boolean have_nmt = true;
+        if (output.getOutput().contains("please enable Native Memory Tracking")) {
+            have_nmt = false;
+        }
```
`boolean have_nmt = output.getOutput().contains....`

jdksjolen commented on src/hotspot/share/services/memMapPrinter.hpp:
```diff
@@ -0,0 +1,66 @@
+  MappingPrintInformation(const void* from, const void* to) : _from(from), _to(to) {}
+  const void* from() const { return _from; }
+  const void* to() const { return _to; }
+  virtual void print_details_1(outputStream* st) const {} // To be printed before VM annotations
+  virtual void print_details_2(outputStream* st) const {} // To be printed before VM annotations
```
May we please have better names :)?

tstuefe replied:
Sure :-)

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+void MemMapPrinter::pd_print_header(outputStream* st) {
+  st->print(
+#ifdef _LP64
+      "from                 to                 "
+#else
+      "from         to         "
```
16+1 and 8+1 spaces, depending on how long the addresses are I assume? Your choice, but might be good to note that.

tstuefe replied:
Ok will add comment

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+void MemMapPrinter::pd_iterate_all_mappings(MappingPrintClosure& closure) {
+  FILE* f = os::fopen("/proc/self/maps", "r");
+  if (f != nullptr) {
+    char line[1024];
```
This looks like a bug: A line may at most be 1024 characters, but `scan_proc_maps_line` may clearly read more than 1024 characters. Yes, it won't read out of bounds, but it limits the parser unnecessarily.

tstuefe replied:
The parser will silently truncate these lines; the rest of the lines will be read with subsequent fgets calls and silently ignored, since they don't match the pattern. This is a pattern that we use all over the place.

About the size, yes, it makes sense to increase this to match what we later read from it.

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+#include "services/memMapPrinter.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+struct proc_maps_info_t {
```
`struct ProcMapsInfo` please. The `_t` is apparently reserved for POSIX, and I prefer our structs and classes to not look like they're coming from a C library anyway.

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+#include "utilities/globalDefinitions.hpp"
+
+struct proc_maps_info_t {
+  unsigned long long from = 0;
+  unsigned long long to = 0;
```
Can we not use `uintptr_t` here? scanf directive: https://stackoverflow.com/questions/12228227/how-to-get-pointer-sized-integer-using-scanf

tstuefe replied:
I'll try void* and %p instead.

jdksjolen reviewed (commented):
Hi,

Thank you for this PR. These comments are just a first pass, I haven't finished going through the code.

tstuefe commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+  char offset[20 + 1];
+  char dev[20 + 1];
+  char inode[20 + 1];
+  char filename[1024 + 1];
```
That is surprisingly tricky since there is no way to specify variable sizes with scanf, unlike printf. 

One could rely on PATH_MAX being a pure numerical ("4096") and assemble the format string with preprocessor magic, but that falls apart if PATH_MAX is e.g. a constant. 

One could build up the format string dynamically at runtime, but that will trigger warnings in code linters and possibly when building since the format string is supposed to be a literal.

Also, the PATH_MAX size is pretty arbitrary. Paths can get infinitely long with nested links. So there is no safe limit anyway. May just as well keep 1024 then, since that should be enough for 99% of cases.

tstuefe commented:
@jdksjolen Many thanks for the review. I fixed most of your requests, and also added a simple timeout fuse to prevent the printing from taking overly long in the rare case of insanely large or fragmented process memory maps.

gerard-ziemski commented:
Taking a look...

gerard-ziemski commented on src/hotspot/share/services/memMapPrinter.cpp:
```diff
@@ -0,0 +1,303 @@
+  st->print_cr("(*) - Mapping contains data from multiple regions");
+  st->cr();
+  pd_print_header(st);
+  // Under rare circumstances the process memory map may be insanely large and/or fragmented. We cap
+  // the absolute runtime of printing to blocking other VM operations too long.
+  const jlong timeout_at = os::javaTimeNanos() +
+                           ((jlong)(SafepointTimeoutDelay * NANOSECS_PER_MILLISEC) / 2);
```
Is AbortVMOnSafepointTimeoutDelay/2 timeout useful for real world cases?

I assume the client who would need this feature, would need it for a reason and that means memory hungry processes?

Instead of blocking VM and having to use the timeout, can we:

A. take a snapshot of NMT virtual memory region and process it without blocking the VM

B. divide this process into smaller chunks, each of which will be guaranteed to finish before AbortVMOnSafepointTimeoutDelay triggers

C. temporarily increase AbortVMOnSafepointTimeoutDelay and revert it back afterwords?

In case A,B we would get back a very close, but not exact view of memory, but wouldn't this be better than getting back precise, but cut off view?

Case C would give us what the users wants assuming they are OK with waiting a bit longer?

And adding malloc'ed memory here would be make this issue much worse.

tstuefe replied:
The whole abort-on-timeout thing I added because I run the command in a VMOp, and it runs in a VMOp because I want to print information about GC threads (so, not because of NMT) and I was not sure how long the GC Thread* objects are guaranteed to live. Otherwise, I would not have bothered. AbortVMOnSafepointTimeoutDelay is by default 10 seconds, that is really plenty.

But.... 

Gaaah! :-( I just tried to dump a million mappings in a test and found that the limiting factor is the output stream size of the jcmd bufferedStream. It looks like jcmd buffers the whole $%!& output before sending it over the wire. For hypothetical situations with millions of mappings, this would mean an additional memory footprint in the GB range. Thats quite annoying.

I opened https://bugs.openjdk.org/browse/JDK-8319055 to track that, but have not high hopes of someone actually having the time for it. A potential solution using temporary buffers (an outputStream variant that under the hood buffers its content in a temp file) would be simple, but a bit embarrassing and may need convincing folks.

I need to think about this...

tstuefe commented:
> I really like this feature, hope the other platforms can be done as well. (I am also really looking forward to seeing how you did the "malloc" version)
> 
> I think we should set the user expectations correctly and say somewhere that these mappings are only for mmaped memory, ex change:
> 
> `Memory mappings:`
> 
> to
> 
> `Memory mappings (mmaped):`
> 
> or something like that.
> 

That would be more confusing since you can see C-heap mappings as well (mappings the libc allocates via mmap to back its arenas). We don't have VM annotations for them (yet), but they do show up in that list.

> These lines look akward next to each other:
> 
> ```
> 0x000000070a400000 - 0x0000000800000000  4123000832  ---p 00000000    JAVAHEAP                      
> 0x00007f4e747dd000 - 0x00007f4e747e1000       16384  ---p 00000000    STACK(72821 "AWT-EventQueue-0") 
> 0x0000563e6decb000 - 0x0000563e6decc000        4096  r--p 00000000                                  /home/gerard/Desktop/Work/8318636/jdk/build/linux-x86_64-server-fastdebug/images/jdk/bin/java
> 0x00007f4e26000000 - 0x00007f4e26d3b000    13873152  rw-p 00001000    CDS                           /home/gerard/Desktop/Work/8318636/jdk/build/linux-x86_64-server-fastdebug/images/jdk/lib/server/classes.jsa
> ```
> 
> in my opinion. The majority of entries either have the component or path in the **info** field (except CDS). Could we print it simply as:
> 
> ```
> 0x000000070a400000 - 0x0000000800000000  4123000832  ---p 00000000    JAVAHEAP                      
> 0x00007f4e747dd000 - 0x00007f4e747e1000       16384  ---p 00000000    STACK(72821 "AWT-EventQueue-0") 
> 0x0000563e6decb000 - 0x0000563e6decc000        4096  r--p 00000000    /home/gerard/Desktop/Work/8318636/jdk/build/linux-x86_64-server-fastdebug/images/jdk/bin/java
> 0x00007f4e26000000 - 0x00007f4e26d3b000    13873152  rw-p 00001000    CDS=/home/gerard/Desktop/Work/8318636/jdk/build/linux-x86_64-server-fastdebug/images/jdk/lib/server/classes.jsa
> ```

Hmm. Yes, that looks better. I was originally worried about VMAs for files being folded with other ones, then you would have confusing entries like "STACK.. filename bla". But that can actually never happen since VMAs with different files (or one has ha file, one don't) are not folded. Yes, I change that.

> 
> I find the compact way more aesthetically pleasing and easier to read.
> 
> Next I'm going to start looking into the implementation...

Thansk!

tstuefe commented:
@gerard-ziemski @jdksjolen I take this back to draft for the moment since I found that I need to solve at least two additional problems. 

One is the time it takes to correlate NMT regions with VMA - its O^2, and experiments with millions of mappings show that this does not scale well. My new plan is to pre-collect all NMT information available first in a flat cache-friendly array that can be iterated cheaply, possibly pre-sorted, and use that when iterating VMAs. And only do the first part in a VM Op.

The second problem is that the output can get huge and I run against limits in dcmd framework. I may just do what other commands do, dump to a file instead to the output.

tstuefe commented on src/hotspot/share/services/memMapPrinter.cpp:
```diff
@@ -0,0 +1,303 @@
+  f(mtInternal,       "INTERN", "Internal") \
+  f(mtJavaHeap,       "JAVAHEAP", "Java Heap") \
+  f(mtOther,          "JDK", "allocated by JDK libraries other than VM") \
+  f(mtMetaspace,      "META", "Metaspace nodes (non-class)") \
+  f(mtSafepoint,      "POLL", "Polling pages") \
+  f(mtThreadStack,    "STACK", "(known) Thread Stack")
+  //end
+
+static const char* get_shortname_for_nmt_flag(MEMFLAGS f) {
+#define DO(flag, shortname, text) if (flag == f) return shortname;
+  NMTFLAGS_DO(DO)
+#undef DO
+  return NMTUtil::flag_to_enum_name(f);
+}
+
+class MemFlagBitmap {
+  uint32_t _v;
+  STATIC_ASSERT(sizeof(_v) * BitsPerByte >= mt_number_of_types);
+
+public:
+  MemFlagBitmap(uint32_t v = 0) : _v(v) {}
+  MemFlagBitmap(const MemFlagBitmap& o) : _v(o._v) {}
+
+  uint32_t raw_value() const { return _v; }
+
+  void set_flag(MEMFLAGS f) {
+    const int bitno = (int)f;
+    _v |= nth_bit(bitno);
+  }
+
+  bool has_flag(MEMFLAGS f) const {
+    const int bitno = (int)f;
+    return _v & nth_bit(bitno);
+  }
+
+  bool has_any() const { return _v > 0; }
+
+};
+
+/// NMT virtual memory
```
Not allocation.hpp, but lets put it into an own header. Arguable the NMT flags should be removed from allocation.hpp and moved somewhere else, since often I find myself including allocation.hpp just for the sake of mt_number_of_flags or some minor thing.

tstuefe commented on src/hotspot/share/services/memMapPrinter.cpp:
```diff
@@ -0,0 +1,303 @@
+
+static uintx safely_get_thread_id(const Thread* t) {
+  const OSThread* osth = t->osthread();
+  uintx tid = 0;
```
I removed the whole function

tstuefe commented on src/hotspot/share/services/memMapPrinter.cpp:
```diff
@@ -0,0 +1,303 @@
+  return 0;
+}
+
+// Given a region [from, to) that is supposed to represent a thread stack,
```
:)

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+
+void MemMapPrinter::pd_iterate_all_mappings(MappingPrintClosure& closure) {
+  FILE* f = os::fopen("/proc/self/maps", "r");
+  if (f != nullptr) {
```
Suggestion: `if (f == nullptr) return;`, reduces indentation for main body of code.

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+  if (f != nullptr) {
+    static constexpr size_t linesize = sizeof(ProcMapsInfo);
+    char line[linesize];
+    while(fgets(line, sizeof(line), f) == line) {
```
Style: space between `while` and `(`.

jdksjolen commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+void MemMapPrinter::pd_iterate_all_mappings(MappingPrintClosure& closure) {
+  FILE* f = os::fopen("/proc/self/maps", "r");
+  if (f != nullptr) {
+    static constexpr size_t linesize = sizeof(ProcMapsInfo);
```
Make this `constexpr const`, remove `static`.

tstuefe replied:
Lets go with constexpr; should imply const.

jdksjolen commented on test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTest.java:
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
```
Copyright Redhat

jdksjolen reviewed (commented):
Hi Thomas,

Did another run through. I haven't looked at `CachedNMTInformation` in depth yet.

gerard-ziemski commented on src/hotspot/share/nmt/memMapPrinter.cpp:
```diff
@@ -0,0 +1,291 @@
+/// NMT virtual memory
+
+static bool range_intersects(const void* from1, const void* to1, const void* from2, const void* to2) {
+  const void* const min = MAX2(from1, from2);
+  const void* const max = MIN2(to1, to2);
+  return min < max;
```
I had to rewrite it as:

`return MAX2(from1, from2) < MIN2(to1, to2);`

to make sure I understand it, or better yet, why not have it as a macros?

`#define RANGE_INTERSECT(min, max) (return MAX2(from1, from2) < MIN2(to1, to2))`

 MAX2 and MIN2 are macros already and we're not doing anything fancy here.

tstuefe replied:
I'll do the former, that is clearer I agree, but leave the latter out (I assume with the macro you mean add it to globalDefenitions.hpp).

I fear that a lot of bikeshedding and general discussions would start, and to do it properly needs a bit more time. Because we really would benefit from having a nice templatized utility classes for ranges. Like MemRegion, but that one is tied to HeapWord* I think.

jdksjolen replied:
`MAX2` and `MIN2` are not macros, they're `constexpr` free functions, and that's what a `RANGE_INTERSECT` free function should be also. Unexpected because of their names, I know.

jdksjolen replied:
FWIW: Yeah, if it's going to be in globalDefinitions.hpp then do that in a separate RFE.

gerard-ziemski commented on src/hotspot/share/nmt/memMapPrinter.cpp:
```diff
@@ -0,0 +1,291 @@
+  struct Range { const void* from; const void* to; };
+  // Unfortunately, we need to allocate manually, raw, since we must prevent
+  // NMT deadlocks (ThreadCritical).
+  Range* _ranges;
+  MEMFLAGS* _flags;
```
Why did you decide to have two separate memory chunks?

I think I'd have used a struct to hold Range* and MEMFLAGS* and use that struct in a single array.

tstuefe replied:
Two reasons:

- more condensed, needs less memory. MEMFLAGS is a byte, Range is 16 bytes. A combined structure would be 24 bytes with padding (since the structure needs to be pointer aligned)
- It is more cache friendly, hence faster. I need to iterate through all ranges, but only need to access the MEMFLAGS for the one range I find. By keeping the ranges condensed, more of them fit into a cache line, so iteration is faster.

gerard-ziemski reviewed (changes requested):
Just a couple of issues/questions.

This is a very nice feature we're adding. Thank you!

tstuefe commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+    static constexpr size_t linesize = sizeof(ProcMapsInfo);
+    char line[linesize];
+    while(fgets(line, sizeof(line), f) == line) {
+      line[sizeof(line) - 1] = '\0';
```
sizeof(line) is the size of the character array, not strlen. It is linesize bytes. `x[sizeof(x) - 1] = '\0';` is the typical pattern one uses to ensure a character array is zero-terminated.

tstuefe commented on src/hotspot/os/linux/memMapPrinter_linux.cpp:
```diff
@@ -0,0 +1,92 @@
+  LinuxMappingPrintInformation(const void* from, const void* to, const ProcMapsInfo* info) :
+    MappingPrintInformation(from, to), _info(*info) {}
+
+  void print_OS_specific_details(outputStream* st) const override {
+    st->print("%s %s ", _info.prot, _info.offset);
+  }
```
This prints protection and offset. The former is interesting for obvious reasons (e.g. lets you tell apart stack guard regions from stack, or uncommitted from committed regions) and the latter interesting for some corner cases (e.g. ZGC mapped regions).

I don't want to do this in shared code though since I am not sure every OS has this information or whether I can obtain this information. I rather not print out a bunch of 0s or empty strings.

The alternative would be carrying a non-descriptive text string around with "OS information", but I don't want to have to think about storage and copying stuff around. It would not be simpler than this.

BTW, the way to implement this on MacOS would be probably by spawning off `vmmap` and parsing the output; on Windows one would use `VirtualQuery`. But I leave this for other folks, I am happy with Linux for now.

bric3 replied:
On ZGC, by the way, it's possible to see many `memfd:java_heap (deleted)`, does this work allow annotate these mappings ?

tstuefe replied:
Yes, they should show up as JAVAHEAP.

jdksjolen commented on src/hotspot/share/nmt/memMapPrinter.cpp:
```diff
@@ -0,0 +1,291 @@
+    ALLOW_C_FUNCTION(free, ::free(_flags);)
+  }
+
+  bool add(const void* from, const void* to, MEMFLAGS f) {
```
Please mention that we're `add`ing in sorted order, that is that `forall R \in _ranges R.to <= from` holds.

tstuefe replied:
I wasn't sure about that. Do we always? Are NMT regions guaranteed to be sorted? (Dimly remember cases where that weren't so)

jdksjolen replied:
It's always sorted in some way, we save the linked lists in some different sorting orders for baselining, but in the `VirtualMemoryTracker` they're guaranteed to be sorted by address.

jdksjolen commented on src/hotspot/share/nmt/memMapPrinter.cpp:
```diff
@@ -0,0 +1,291 @@
+
+  bool add(const void* from, const void* to, MEMFLAGS f) {
+    assert(_count == 0 || (from >= _ranges[_count - 1].to), "NMT regions unordered?");
+    // we can just fold two regions if they are adjacent and have the same flag.
+    if (_count > 0 && from == _ranges[_count - 1].to && f == _flags[_count - 1]) {
+      _ranges[_count - 1].to = to;
+      return true;
+    }
```
I'm pretty sure that the virtual memory tracker already gives you the minimal set of regions, have you observed this branch being taken?

tstuefe replied:
Yes, it resulted in a significant performance gain for a test run where I interleaved non-committed and committed memory of the same tag. I may have been mistaken, of course.

jdksjolen replied:
It being committed or reserved shouldn't matter (I assume non-committed = reserved), as committed just means that there's a pointer to the committed memory region in the reserved memory region. Well, all I can say is that I'm surprised.

tstuefe replied:
They differ by permissions. I guess NMT keeps them separate for that reason. In any case, the code snippet makes a huge difference.

jdksjolen commented on src/hotspot/share/nmt/memMapPrinter.cpp:
```diff
@@ -0,0 +1,291 @@
+      _capacity = new_capacity;
+    }
+    assert(_capacity > _count, "Sanity");
+    _ranges[_count].from = from;
+    _ranges[_count].to = to;
```
Or just `_ranges[_count] = Range{from, to}`

tstuefe replied:
ok

jdksjolen reviewed (commented):
OK, went through the cache. Will continue later.

gerard-ziemski reviewed (approved):
LGTM

Thank you for doing this!

tstuefe commented:
> LGTM
> 
> Thank you for doing this!

Thank you for reviewing, @gerard-ziemski !

tstuefe commented:
Windows GHA errors unrelated.
@jdksjolen are we good to go?

tstuefe commented:
@jdksjolen anything more needed from my side?

jdksjolen commented:
> @jdksjolen anything more needed from my side?

I'll work through this tomorrow, just to make sure I didn't miss anything.

jdksjolen commented:
As this adds a JCmd, doesn't this need both a CSR and a manual entry?

tstuefe commented:
> As this adds a JCmd, doesn't this need both a CSR and a manual entry?

- CSR: not sure; there are precedences for going with CSR and without CSR. Since we get awfully close to JDK22 freeze, I would prefer for a CSR not to be needed. Also would make backporting a lot easier.
- Manpage: not sure either; IIRC last time I tried, I was told that Oracle maintains these internally, because there is internal documentation that needs updating too?

jdksjolen commented:
> > As this adds a JCmd, doesn't this need both a CSR and a manual entry?
> 
>     * CSR: not sure; there are precedences for going with CSR and without CSR. Since we get awfully close to JDK22 freeze, I would prefer for a CSR not to be needed. Also would make backporting a lot easier.
> 
>     * Manpage: not sure either; IIRC last time I tried, I was told that Oracle maintains these internally, because there is internal documentation that needs updating too?

@dholmes-ora, would you mind helping us out here with regards to the process? Would a new JCmd require a CSR or would it be acceptable to merge without one? Thank you.

dholmes-ora commented:
> > > As this adds a JCmd, doesn't this need both a CSR and a manual entry?
> > 
> > 
> > ```
> > * CSR: not sure; there are precedences for going with CSR and without CSR. Since we get awfully close to JDK22 freeze, I would prefer for a CSR not to be needed. Also would make backporting a lot easier.
> > 
> > * Manpage: not sure either; IIRC last time I tried, I was told that Oracle maintains these internally, because there is internal documentation that needs updating too?
> > ```
> 
> @dholmes-ora, would you mind helping us out here with regards to the process? Would a new JCmd require a CSR or would it be acceptable to merge without one? Thank you.

@jdksjolen No CSR needed: from another related PR - "We do not use CSR requests with jcmd changes as they are deemed diagnostic commands - ref JDK-8203682"

But yes the `jcmd` manpage should be updated ref:
https://docs.oracle.com/en/java/javase/19/docs/specs/man/jcmd.html
though I'm worried we may not have kept it up to date! That requires an Oracle engineer to apply the changes to the jcmd.md markdown sources in our repo, and then regenerate the `jcmd.1` manpage file. The doc update can be split into a separate doc sub-task so that the main PR is not held up. (And we probably need an audit to see if any other updates are missing - which is painful.)

dholmes-ora commented on src/hotspot/share/services/diagnosticCommand.hpp:
```diff
@@ -978,4 +978,41 @@ class CompilationMemoryStatisticDCmd: public DCmdWithParser {
+  SystemMapDCmd(outputStream* output, bool heap);
+  static const char* name() { return "System.map"; }
+  static const char* description() {
+    return "Prints an annotated process memory map of the VM process.";
```
Should this specify "Linux Only" ?

tstuefe commented:
> > > > As this adds a JCmd, doesn't this need both a CSR and a manual entry?
> > > 
> > > 
> > > ```
> > > * CSR: not sure; there are precedences for going with CSR and without CSR. Since we get awfully close to JDK22 freeze, I would prefer for a CSR not to be needed. Also would make backporting a lot easier.
> > > 
> > > * Manpage: not sure either; IIRC last time I tried, I was told that Oracle maintains these internally, because there is internal documentation that needs updating too?
> > > ```
> > 
> > 
> > @dholmes-ora, would you mind helping us out here with regards to the process? Would a new JCmd require a CSR or would it be acceptable to merge without one? Thank you.
> 
> @jdksjolen No CSR needed: from another related PR - "We do not use CSR requests with jcmd changes as they are deemed diagnostic commands - ref JDK-8203682"
> 
> But yes the `jcmd` manpage should be updated ref: https://docs.oracle.com/en/java/javase/19/docs/specs/man/jcmd.html though I'm worried we may not have kept it up to date! That requires an Oracle engineer to apply the changes to the jcmd.md markdown sources in our repo, and then regenerate the `jcmd.1` manpage file. The doc update can be split into a separate doc sub-task so that the main PR is not held up. (And we probably need an audit to see if any other updates are missing - which is painful.)

Would be nice if that process were open like the rest. It would take load from Oracle and most developers would be fine with maintaining the man page themselves.

dholmes-ora commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -471,6 +471,7 @@ void Thread::print_on(outputStream* st, bool print_extended_info) const {
     if (!is_Java_thread() || !JavaThread::cast(this)->is_vthread_mounted()) {
       osthread()->print_on(st);
     }
+    st->print("stack=[" PTR_FORMAT ", " PTR_FORMAT ")", p2i(stack_end()), p2i(stack_base()));
```
We already report the thread stack elsewhere, so is this going to duplicate the information?

dholmes-ora commented on src/hotspot/share/services/diagnosticCommand.hpp:
```diff
@@ -978,4 +978,41 @@ class CompilationMemoryStatisticDCmd: public DCmdWithParser {
+  SystemDumpMapDCmd(outputStream* output, bool heap);
+  static const char* name() { return "System.dump_map"; }
+  static const char* description() {
+    return "Dumps an annotated process memory map to an output file.";
```
Again "Linux Only"?

dholmes-ora commented on src/hotspot/share/runtime/vmOperation.hpp:
```diff
@@ -117,7 +117,8 @@
   template(JvmtiPostObjectFree)                   \
-  template(RendezvousGCThreads)
+  template(RendezvousGCThreads)                   \
+  template(SystemMap)
```
I can't spot the VMop for this. ??

dholmes-ora commented on src/hotspot/share/services/diagnosticCommand.cpp:
```diff
@@ -1151,3 +1155,41 @@ void CompilationMemoryStatisticDCmd::execute(DCmdSource source, TRAPS) {
   CompilationMemoryStatistic::print_all_by_size(output(), human_readable, minsize);
 }
+
+SystemMapDCmd::SystemMapDCmd(outputStream* output, bool heap) :
```
Shouldn't this all be inside `ifdef LINUX`? You don't need `os::realpath` if you do that.

dholmes-ora replied:
Well to call `os::Posix::realpath` it would need Linux specific includes ... but as it is you have unused code in the Windows version.

dholmes-ora commented on src/hotspot/share/services/diagnosticCommand.cpp:
```diff
@@ -1151,3 +1155,41 @@ void CompilationMemoryStatisticDCmd::execute(DCmdSource source, TRAPS) {
+    name = absname != nullptr ? absname : name;
+    output()->print_cr("Memory map dumped to \"%s\".", name);
+  } else {
+    output()->print_cr("Failed to open \"%s\" for writing.", name);
```
Don't you want to report errno here?

tstuefe replied:
I'm always leery about relying on errno from wrapper functions or objects like this, unless they have an explicit channel to report the relevant errno. I looked at fileStream, I guess its okay here. Okay, I'll add it.

dholmes-ora reviewed (changes requested):
If the command is only usable on Linux, and only registered on Linux, do we really need the empty definitions in `memMapPrinter_<os>.cpp`? At present it seems we only need it because all the implementation code is not isolated to `ifdef LINUX`.

tstuefe commented:
@dholmes-ora thank you for your review. I hope I addressed all concerns. I made this all Linux only, removed os::realpath, print errno.

I am currently working on a followup RFE that provides this command for Windows, and it works, but I will do it in a separate RFE (rolling back the ifdef LINUX should be easy) and possibly not for JDK 22. https://github.com/openjdk/jdk/pull/16593

jdksjolen reviewed (approved):
Hi Thomas,

Thanks for your hard work on this! LGTM.

tstuefe commented:
> Hi Thomas,
> 
> Thanks for your hard work on this! LGTM.

Thank you for your Review, Johan!

dholmes-ora reviewed (commented):
Thanks for the updates @tstuefe  - nothing further from me (with the Windows variant in the pipeline I will wait and see how things look in terms of shared versus platform-specific code.

Please create a JBS issue to get this new command documented in the jcmd manpage, with details on what that documentation should look like.

Thanks

tstuefe commented:
Thanks @dholmes-ora !

/integrate


> Thanks for the updates @tstuefe - nothing further from me (with the Windows variant in the pipeline I will wait and see how things look in terms of shared versus platform-specific code.
> 
> Please create a JBS issue to get this new command documented in the jcmd manpage, with details on what that documentation should look like.

There are a number of inconsistencies apart from this new dcmd; I think the jcmd man page needs to be updated.

I collected all inconsistencies I could spot in this issue: https://bugs.openjdk.org/browse/JDK-8319948 and put it onto your name.

> 
> Thanks

Thank you, David.
